### ISSUE TITLE
tracker improvements

### ISSUE THREAD

**Issue by psenough** [nesbox/TIC-80#1404]
Hi,

Was trying the tracker in TIC-80 recently (v0.80.1344) for a collaboration with jeenio, and we found a few things that could help make the experience more intuitive to pick up for newcomers and seasoned tracker software users alike, most are just general quality of life suggestions. Here is a dump of that list:

- track is called song on other trackers
- sfx are called instruments in other trackers
- frames are called sequence patterns in other trackers
- in the sfx editor it's hard to intuitively understand that you're editing the sfx index from the top left. it feels like the sfx are the waveforms instead. maybe have a tighter gui connection between the top left and the right side of the screen. so the waveforms would feel more detached from that section.
- i also end up clicking the waveforms to look at their form in higher res and don't imediately realize i am filling out the selected sfx with that wave. maybe a button to "apply this wave on whole sfx" would make more sense and let people edit/preview the waves without affecting the sfx currently selected.
- missing a way to define channel volume / panning
- would be nice if the shortcuts were the same as some vintage trackers 
:: reference of impulse tracker / schism tracker shortcuts
f5 play whole song
f6 play current pattern
f7 play from mark / current row
f8 stop playback
* increase octave
/ lower octave
:: reference of fasttracker 2 / renoise shortcuts
Space: Start/Stop playing.
Enter: Only play the line that the cursor is on (play step by step).
Right Alt: Start playing and looping the current pattern.
Right Ctrl: Start playing the song.

jeenio's feedback of what felt missing/lacking:
- possibility to define per channel initial volume/panning (so it doesn't take effects column space at the start. maybe even define per frame?)
- shuffle editor: either by percentage (66%, for instance, would be triplet shuffle)
- row highlight (4, 8, and 16 rows, maybe different shades, maybe customizable or simply 1/2 - 1/4 - 1/8 - 1/16 of total number of rows?) and yes, I've noticed the dots... but only today! :slight_smile:
- store sfx/instruments on disc (so they can be reused in later songs, to prevent to have to either copy/paste between projects or redo them by hand)
- it happened to me that I would start a new frame, reuse an old pattern ID, start to change it and only after some time realize that I ruined a previous frame because I forgot to copy the pattern ID to a new one. I don't have an elegant suggestion, as I'm not fan of dialog boxes popping up saying "This pattern is being used in another frame. Are you sure you want to modify it?" but something should remind me that is being used elsewhere. maybe a color change in the pattern ID? like, it could be white if unique and if being used in another pattern change to yellow....?
- different behaviour of delete and backspace keys. right now, delete key just deletes the note, but it would be fundamental to have a delete option that would move the contents below one row up, just like text editors, when you delete a character the rest backs up and doesn't leave a space.
- selecting only works per channel. would be nice to be able to select across channels, though it could be easily a mess with unique pattern IDs... but imagine you have one frame with pattern IDs 01, 02, 03 and 04 and you want to create another frame with those patterns but with small changes. would be nice to just copy those patterns to 05, 06, 07 and 08 or part of them with the selection tool.
- this one is a long shot: would be nice to be able to view more rows at once. this probably goes against the concept of the whole thing...

thanks for your work on the platform!

**Comment by nesbox** [comment]
Awesome, thank you for the feedback, will try to improve it...

**Comment by joshgoebel** [comment]
Please @ me if you make changes here as a lot of this probably will likely touch the wiki as well.

**Comment by nesbox** [comment]
@joshgoebel sure, thank you

**Comment by Kyuchumimo** [comment]
It is assumed that you can define the master volume and panning with the command Mxy, where x and y are a 4-bit value from 0 to F, x is the left channel and y is the right channel

**Comment by psenough** [comment]
> It is assumed that you can define the master volume and panning of a single channel with the command Mxy, where x and y are a 4-bit value from 0 to F, x is the left channel and y is the right channel

yeah but if you wan't to play a note with a chord (or some other effect) you can't set the volume at the same time, so you end up needing to use tricks (like playing a 1 row pattern just to set the volumes) would be nice to have a cleaner way to set configure the default global volume of the pattern that would free up the effects columns for other actual effects.

**Comment by joshgoebel** [comment]
I guess changing the volume constantly isn't perhaps a common use case?

**Comment by Kyuchumimo** [comment]
> > It is assumed that you can define the master volume and panning of a single channel with the command Mxy, where x and y are a 4-bit value from 0 to F, x is the left channel and y is the right channel
> 
> yeah but if you wan't to play a note with a chord (or some other effect) you can't set the volume at the same time, so you end up needing to use tricks (like playing a 1 row pattern just to set the volumes). would be nice to have a cleaner way to set configure the default global volume of the pattern/channel that would free up the effects columns for other actual effects.

you can use arpeggios within the sfx editor to replace the chord effect

**Comment by farvardin** [comment]
I'm playing a bit with the music tracker, and so far what I miss the most is the hability to start a track from the cursor, instead of playing it from the start...

**Comment by nesbox** [comment]
@farvardin try `shift+enter` to start playing from the cursor

**Comment by jeenio** [comment]
hey @nesbox, great improvement already! a few issues though:

- ctrl+up/down to change instruments is great, but when at 00 you can go to 63, but not the other way round
- ctrl+left/right to change frames is a life saver, but would be nice to be able to wrap from 15 to 00 and vice versa
- changing instruments and transposing (with ctrl+F1/F2/F3/F4) should actually be audible
- suggestion: to select all notes in all channels - ctrl + aa (first "a" press selects the channel, second "a" press selects all channels)
- suggestion: ctrl+1/2/3/4 could change the on/off channel state
- suggestion: we have ctrl+f for follow, why not ctrl+l for loop?
- suggestion: pressing 8 would play current row. great to preview chords
- on my laptop, page up/down (shift + 9/3 numpad) doesn't work, while in any other program system wide I have no issue. home and end the same problem.

and thanks for all the work!
